infra/cifuzz/filestore/git/__init__.py

```diff
@@ -0,0 +1,155 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Module for a git based filestore."""
+
+from distutils.dir_util import copy_tree
+import logging
+import os
+import shutil
+import subprocess
+import tempfile
+
+import filestore
+
+_PUSH_RETRIES = 3
+_GIT_EMAIL = 'cifuzz@clusterfuzz.com'
+_GIT_NAME = 'CIFuzz'
+_CORPUS_DIR = 'corpus'
+_COVERAGE_DIR = 'coverage'
+
+
+def git_runner(repo_path):
+  """Returns a gits runner for the repo_path."""
+
+  def func(*args):
+    return subprocess.check_call(('git', '-C', repo_path) + args)
+
+  return func
+
+
+# pylint: disable=unused-argument,no-self-use
+class GitFilestore(filestore.BaseFilestore):
+  """Generic git filestore. This still relies on another filestore provided by
+  the CI for larger artifacts or artifacts which make sense to be included as
```

Comment:
I don't really understand why larger artifacts shouldn't be part of git uploads. Can you explain?

Reply:
> > I did a first pass and left some comments to understand things at a high-level. I'll do a more detailed review of the git parts later today.
> > I understand why crashes need to be uploaded with the artifact filestore. But I'm less clear on the builds part. What does supporting two filestores give us that just one doesn't. As far as I can tell the problems with each build individually are:
> > artifacts:
> > 
> > 1. Inelegant and hacky.
> > 2. Artifacts are not browseable.
> > 
> > git:
> > 
> > 1. Can't handle large uploads?
> > 2. Can't display things to users?
> 
> The main reason is that git isn't designed for large uploads. There's git LFS but github's quota is limited: https://docs.github.com/en/github/managing-large-files/versioning-large-files/about-storage-and-bandwidth-usage


I've heard this before, but what are the consequences of ignoring this?
I know that deleting binary files doesn't really reduce repo size, but I think this shouldn't be a problem if this repo is only used for CIFuzz purposes right?

> 
> > FYI I made a change (accidentally pushed to master [7817498](https://github.com/google/oss-fuzz/commit/781749884b4ae1446b4e4e9f626524f090889520) that adds raw_upload_directory and raw_download_artifact functions. You can use this for uploading the crashes without the internal tarball.
> 
> Awesome. I don't think there's anything additional I need to do in this PR right? we're just piggy backing on _upload_directory, where you made this change.

There is a change you need to make to take advantage of this. In this PR you should change the typed upload_crashes function to use raw_upload_directory instead of _upload_directory which tars it first.